src/components/forms/DiskDeviceFormCustom.tsx

```diff
@@ -32,38 +35,38 @@ interface Props {
 
 const DiskDeviceFormCustom: FC<Props> = ({ formik, project, profiles }) => {
   const readOnly = (formik.values as EditInstanceFormValues).readOnly;
-  const customVolumes = formik.values.devices
-    .filter((item) => item.type === "disk" && !isRootDisk(item))
-    .map((device) => device as FormDiskDevice);
-
   const existingDeviceNames = getExistingDeviceNames(formik.values, profiles);
 
-  const addVolume = (volume: LxdStorageVolume) => {
+  const addDiskDevice = (device: LxdDiskDevice) => {
     const copy = [...formik.values.devices];
-    copy.push({
-      type: "disk",
-      name: deduplicateName("volume", 1, existingDeviceNames),
-      path: volume.content_type === "filesystem" ? "" : undefined,
-      pool: volume.pool,
-      source: volume.name,
-    });
+    const newDevice: FormDevice = {
+      ...device,
+      name: deduplicateName("disk-device", 1, existingDeviceNames),
+    };
+
+    copy.push(newDevice);
     void formik.setFieldValue("devices", copy);
 
     const name = `devices.${copy.length - 1}.path`;
     focusField(name);
   };
 
-  const changeVolume = (
-    volume: LxdStorageVolume,
-    formVolume: FormDiskDevice,
+  const changeDiskDevice = (
+    newDevice: LxdDiskDevice,
+    existingDevice: FormDiskDevice,
     index: number,
   ) => {
-    void formik.setFieldValue(`devices.${index}.pool`, volume.pool);
-    void formik.setFieldValue(`devices.${index}.source`, volume.name);
-    if (volume.content_type === "filesystem" && formVolume.path === undefined) {
+    void formik.setFieldValue(`devices.${index}.pool`, newDevice.pool);
+    void formik.setFieldValue(`devices.${index}.source`, newDevice.source);
+
+    // If path must exist for the device, and it's not set for the existing device, set it to an empty string
+    // otherwise keep the existing value
+    if (newDevice.path === "" && existingDevice.path === undefined) {
       void formik.setFieldValue(`devices.${index}.path`, "");
     }
-    if (volume.content_type === "block") {
+
+    // If path must not exist for the device, remote it even if it's set for the existing device
+    if (newDevice.path === undefined) {
       void formik.setFieldValue(`devices.${index}.path`, undefined);
     }
```

Comment:
This is a bit confusing. Can you describe in the comments why this is needed and avoid describing what is happening? I think for the what, the code is easy enough to gasp it with a glance.

Reply:
Yeah looking at it again, the fact that path === "" means storage volume has file system content type is completely lost here. Only way is to move the device translation up one more layer

Reply:
done, this function is basically reverted to its original state